src/registry/domain/components-details.js

```diff
@@ -0,0 +1,73 @@
+'use strict';
+
+const async = require('async');
+const _ = require('lodash');
+
+const eventsHandler = require('./events-handler');
+const getUnixUTCTimestamp = require('../../utils/get-unix-utc-timestamp');
+
+module.exports = (conf, cdn) => {
+
+  const returnError = (code, message, callback) => {
+    eventsHandler.fire('error', { code, message });
+    return callback(code);
+  };
+
+  const filePath = () => `${conf.s3.componentsDir}/components-details.json`;
+
+  const getFromJson = callback => cdn.getJson(filePath(), true, callback);
+
+  const getFromDirectories = (options, callback) => {
+
+    const details = _.extend({}, _.cloneDeep(options.details));
+    details.components = details.components || {};
+
+    async.eachOfSeries(options.componentsList.components, (versions, name, done) => {
+
+      details.components[name] = details.components[name] || {};
+
+      async.eachLimit(versions, cdn.maxConcurrentRequests, (version, next) => {
+        if(details.components[name][version]){
+          next();
+        } else {
+          cdn.getJson(`${conf.s3.componentsDir}/${name}/${version}/package.json`, true, (err, content) => {
+            if(err){ return next(err); }
+            details.components[name][version] = { publishDate: content.oc.date || 0 };
+            next();
+          });
+        }
+      }, done);
+    }, (err) => callback(err, {
+      lastEdit: getUnixUTCTimestamp(),
+      components: details.components
+    }));
+  };
+
+  const save = (data, callback) => cdn.putFileContent(JSON.stringify(data), filePath(), true, callback);
+
+  const refresh = (componentsList, callback) => {
```

Comment:
The refresh signature (called on registry start and component published) has the components list (all the components and all versions) as input. The reason is that both times this is needed to happen, it happens after the list is fetched or modified, so it is good to start from there for navigating the tree.